test/test_markets.js

```diff
@@ -0,0 +1,94 @@
+const utils = require('./utils')
+
+const Event = artifacts.require('Event')
+const EventFactory = artifacts.require('EventFactory')
+const Token = artifacts.require('Token')
+const EtherToken = artifacts.require('EtherToken')
+const CentralizedOracle = artifacts.require('CentralizedOracle')
+const CentralizedOracleFactory = artifacts.require('CentralizedOracleFactory')
+const Market = artifacts.require('Market')
+const StandardMarketFactory = artifacts.require('StandardMarketFactory')
+const LMSRMarketMaker = artifacts.require('LMSRMarketMaker')
+
+contract('markets', function (accounts) {
+    let centralizedOracleFactory
+    let eventFactory
+    let etherToken
+    let standardMarketFactory
+    let lmsrMarketMaker
+
+    beforeEach(async () => {
+        centralizedOracleFactory = await CentralizedOracleFactory.deployed()
+        eventFactory = await EventFactory.deployed()
+        etherToken = await EtherToken.deployed()
+        standardMarketFactory = await StandardMarketFactory.deployed()
+        lmsrMarketMaker = await LMSRMarketMaker.deployed()
+    })
+
+    it('should buy and sell all outcomes', async () => {
+        // create event
+        const ipfsHash = 'QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG'
+        const oracle = utils.getParamFromTxEvent(
+            await centralizedOracleFactory.createCentralizedOracle(ipfsHash),
+            'centralizedOracle', CentralizedOracle
+        )
+        const event = utils.getParamFromTxEvent(
+            await eventFactory.createCategoricalEvent(etherToken.address, oracle.address, 2),
+            'categoricalEvent', Event
+        )
+
+        // create market
+        const feeFactor = 50000  // 5%
+        const market = utils.getParamFromTxEvent(
+            await standardMarketFactory.createMarket(event.address, lmsrMarketMaker.address, feeFactor),
+            'market', Market
+        )
+
+        // Fund market
+        const investor = 0
+        const funding = 1e18
+
+        await etherToken.deposit({ value: funding, from: accounts[investor] })
+        assert.equal(await etherToken.balanceOf(accounts[investor]), funding)
+
+        await etherToken.approve(market.address, funding, { from: accounts[investor] })
+
+        await market.fund(funding, { from: accounts[investor] })
+        assert.equal(await etherToken.balanceOf(accounts[investor]), 0)
+
+        // Buy outcome tokens
+        const buyer = 1
+        const outcome = 0
+        const token_count = 1e15
+        const outcome_token_cost = await lmsrMarketMaker.calcCost(market.address, outcome, token_count)
```

Comment:
All numbers that come back from contract calls are BigNumbers. I've discovered that you can use them in `assert.equal` calls directly as long as _both_ args are not `BigNumber`. You can convert them to plain JS numbers by using the `.valueOf()` method, or by implicitly casting them via _some_ arithmetic ops. For some reason, if you try to use `+` with `BigNumber` and native numbers, you will end up concatenating their decimal representations. WTF.

Anyway, those are my caveats with that type.